### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by TurtleMinecraft**

### CODE REVIEW

File: .idea/uiDesigner.xml [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+      </item>
+      <item class="javax.swing.JToolBar$Separator" icon="/com/intellij/uiDesigner/icons/toolbarSeparator.svg" removable="false" auto-create-binding="false" can-attach-label="false">
+        <default-constraints vsize-policy="0" hsize-policy="0" anchor="0" fill="1" />
+      </item>
+      <item class="javax.swing.JScrollBar" icon="/com/intellij/uiDesigner/icons/scrollbar.svg" removable="false" auto-create-binding="true" can-attach-label="false">
+        <default-constraints vsize-policy="6" hsize-policy="0" anchor="0" fill="2" />
+      </item>
+    </group>
+  </component>
+</project>
```
Comment (tuval88) [inline .idea/uiDesigner.xml]: No empty line at end of file

File: src/main/java/frc/robot/commands/Shoot.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+public class Shoot extends ParallelDeadlineGroup {
+
+    public static final RootNamespace ROOT = new RootNamespace("shoot");
+
+    private static final SpikesLogger LOGGER = new SpikesLogger();
+
+    public static final double CLOSE_HEIGHT = 24.6;
+    public static final double MIDDLE_HEIGHT = 22;
+    public static final double RECTANGLE_HEIGHT = 16.3;
```
Comment (tuval88) [inline src/main/java/frc/robot/commands/Shoot.java]: These require documentation, and possibly better names.

File: src/main/java/frc/robot/commands/Shoot.java [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+public class Shoot extends ParallelDeadlineGroup {
+
+    public static final RootNamespace ROOT = new RootNamespace("shoot");
+
+    private static final SpikesLogger LOGGER = new SpikesLogger();
```
Comment (tuval88) [inline src/main/java/frc/robot/commands/Shoot.java]: reorder

File: src/main/java/frc/robot/OI.java [hunk 4]
```diff
@@ -1,8 +1,67 @@
...
+        xbox.getGreenButton().onTrue(new IntakeNote(intakeRoller, storage, intakePlacer, shooterAdjuster, false));
+        // Shoots the note from the subwoofer - B
+        xbox.getRedButton().onTrue(new Shoot(shooter, drivetrain, shooterAdjuster, storage,
+                Shoot.CLOSE_HEIGHT).getCommand());
+        // Shoots the note from the safe zone - X
+        xbox.getBlueButton().onTrue(new Shoot(shooter, drivetrain, shooterAdjuster, storage,
+                Shoot.RECTANGLE_HEIGHT).getCommand());
+        // Moves the storage backwards - LT
+        xbox.getLTButton().whileTrue(new MoveGenericSubsystem(storage, STORAGE_RETRACTION_SPEED));
+        // Retracts the shooter's flywheels - RT
```
Comment (tuval88) [inline src/main/java/frc/robot/OI.java]: Reword to make consistent with the comment above

File: src/main/java/frc/robot/Robot.java [hunk 5]
```diff
@@ -4,68 +4,150 @@
...
-
-  }
-
-  @Override
-  public void simulationPeriodic() {
-
-  }
+    private final RootNamespace root = new RootNamespace("robot");
+    private final SendableChooser<Command> autoChooser = new SendableChooser<>();
+
```
Comment (tuval88) [inline src/main/java/frc/robot/Robot.java]: Why is this here?

File: src/main/java/frc/robot/Robot.java [hunk 6]
```diff
@@ -4,68 +4,150 @@
...
+        registerPathPlannerCommands();
+        configureDashboard();
+        configureAutoChooser();
+    }
+
+    @Override
+    public void robotPeriodic() {
+        logger.log(DriverStation.getAlliance().get() == DriverStation.Alliance.Red ? "red" : "blue");
+        led.periodic();
+        led.preGame();
```
Comment (tuval88) [inline src/main/java/frc/robot/Robot.java]: Move this to DisabledPeriodic

File: src/main/java/frc/robot/RobotMap.java [hunk 7]
```diff
@@ -4,21 +4,59 @@ public class RobotMap {
...
+        int RIGHT_SHOOTER_SPARK_MAX = 14;
+
+        int SHOOTER_ADJUSTER_SPARK_MAX = 15;
+
+        int INTAKE_ROLLER_SPARK_MAX = 19;
+
+        int INTAKE_PLACER_LEFT_SPARK_MAX = 17;
+        int INTAKE_PLACER_RIGHT_SPARK_MAX = 18;
+
+        int STORAGE_SPARK_MAX = 16;
```
Comment (tuval88) [inline src/main/java/frc/robot/RobotMap.java]: Reorder CAN ports to be in numerical order

File: src/main/java/frc/robot/commands/Adjust.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
+package frc.robot.commands;
+
+import com.spikes2212.command.genericsubsystem.commands.smartmotorcontrollergenericsubsystem.MoveSmartMotorControllerGenericSubsystem;
```
Comment (tuval88) [inline src/main/java/frc/robot/commands/Adjust.java]: *ify

File: src/main/java/frc/robot/commands/Adjust.java [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+
+    public Adjust(ShooterAdjuster adjuster, Supplier<Double> setpoint) {
+        addRequirements(adjuster);
+
+        if (Math.abs(adjuster.getPosition() - setpoint.get()) > PID_START_POINT) {
+            addCommands(new FunctionalCommand(() -> {
+                    },
+                            () -> adjuster.set(SPEED * Math.signum(adjuster.getPosition() - setpoint.get())),
+                            interrupted -> adjuster.stop(),
+                            () -> Math.abs(adjuster.getPosition() - setpoint.get()) > PID_START_POINT),
```
Comment (tuval88) [inline src/main/java/frc/robot/commands/Adjust.java]: Is this condition not inverted?

File: src/main/java/frc/robot/commands/Adjust.java [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+
+    public final RootNamespace namespace = new RootNamespace("adjust");
+
+    public Adjust(ShooterAdjuster adjuster, Supplier<Double> setpoint) {
+        addRequirements(adjuster);
+
+        if (Math.abs(adjuster.getPosition() - setpoint.get()) > PID_START_POINT) {
+            addCommands(new FunctionalCommand(() -> {
+                    },
+                            () -> adjuster.set(SPEED * Math.signum(adjuster.getPosition() - setpoint.get())),
```
Comment (tuval88) [inline src/main/java/frc/robot/commands/Adjust.java:29]: What even is this formatting?

File: src/main/java/frc/robot/commands/DriveSwerve.java [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+    private static final double DRIVE_ACCELERATION_LIMIT = 2;
+    private static final double TURN_ACCELERATION_LIMIT = 4;
+
+    private final Drivetrain drivetrain;
+    private final Supplier<Double> xSpeed;
+    private final Supplier<Double> ySpeed;
+    private final Supplier<Double> rotationSpeed;
+    private final boolean fieldRelative;
+    private final boolean usePID;
+    private final boolean useMinSpeed;
```
Comment (tuval88) [inline src/main/java/frc/robot/commands/DriveSwerve.java:24]: Document

File: src/main/java/frc/robot/commands/DriveSwerve.java [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+        if (useMinSpeed) {
+            if (Math.abs(xSpeed) < Drivetrain.MIN_SPEED_METERS_PER_SECONDS) xSpeed = 0;
+            if (Math.abs(ySpeed) < Drivetrain.MIN_SPEED_METERS_PER_SECONDS) ySpeed = 0;
+            if (Math.abs(rotationSpeed) < Drivetrain.MIN_SPEED_METERS_PER_SECONDS) rotationSpeed = 0;
+        }
+        drivetrain.drive(xSpeed, ySpeed, rotationSpeed, fieldRelative, usePID);
+    }
+
+    @Override
+
```
Comment (tuval88) [inline src/main/java/frc/robot/commands/DriveSwerve.java]: Delete this. Now.

File: src/main/java/frc/robot/commands/Shoot.java [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+            Optional<DriverStation.Alliance> alliance = DriverStation.getAlliance();
+            if (alliance.isEmpty() || alliance.get() == DriverStation.Alliance.Blue) {
+                speakerPose = BLUE_SPEAKER_POSE;
+            } else {
+                speakerPose = RED_SPEAKER_POSE;
+            }
+        });
+
+        addCommands(setSpeakerPoseCommand, speedUpCommand, adjustCommand);
+        ledService.attemptShoot();
```
Comment (tuval88) [inline src/main/java/frc/robot/commands/Shoot.java:91]: Is this supposed to just be running during command construction?

File: src/main/java/frc/robot/commands/ShootWithParameters.java [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+        RotateSwerveWithPID rotateCommand = new RotateSwerveWithPID(drivetrain, requiredAngle,
+                drivetrain::getAngle, drivetrain.rotateToTargetPIDSettings,
+                drivetrain.rotateToTargetFeedForwardSettings);
+
+        SpeedUpShooter speedUpCommand = new SpeedUpShooter(shooter, requiredLeftSpeed,
+                requiredRightSpeed);
+
+        MoveSmartMotorControllerGenericSubsystem adjustCommand =
+                new MoveSmartMotorControllerGenericSubsystem(adjuster, adjuster.getPIDSettings(),
+                        adjuster.getFeedForwardSettings(), UnifiedControlMode.POSITION, requiredHeight);
```
Comment (tuval88) [inline src/main/java/frc/robot/commands/ShootWithParameters.java]: isFinished does not check Adjuster.wasReset()

File: src/main/java/frc/robot/commands/auto/YeetAndRetreatAmpSide.java [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+                new InstantCommand(intakePlacer::resetPosition),
+                adjuster.getResetCommand(),
+                new Shoot(shooter, drivetrain, adjuster, storage, Shoot.CLOSE_HEIGHT).getCommand(),
+                new DriveSwerve(drivetrain, DRIVE_SPEED, STRAFE_SPEED, () -> 0.0, false,
+                        false).withTimeout(MOVE_TIMEOUT)
+        );
+    }
+
+    // depending on the alliance, the strafe direction changes as the amps are on opposite sides
+    private static int getStrafeSign() {
```
Comment (tuval88) [inline src/main/java/frc/robot/commands/auto/YeetAndRetreatAmpSide.java]: Why does this function not just return a `double` if its return value is only ever used as one?

File: src/main/java/frc/robot/services/poseestimation/PhotonCameraEstimator.java [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+ * Class which represents a camera running photonvision which is used to get the robot pose.
+ */
+public class PhotonCameraEstimator extends BasePoseSource {
+
+    private final PhotonCamera photonCamera;
+    private final PhotonPoseEstimator poseEstimator;
+    private final Transform3d cameraToRobotCenter;
+    private SpikesLogger logger = new SpikesLogger();
+
+    public PhotonCameraEstimator(String cameraName, Transform3d cameraToRobotCenter) {
```
Comment (tuval88) [inline src/main/java/frc/robot/services/poseestimation/PhotonCameraEstimator.java]: These parameters should be reordered as per Convention A-III\b3

File: src/main/java/frc/robot/services/poseestimation/PhotonCameraEstimator.java [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+        super(cameraToRobotCenter);
+        this.photonCamera = new PhotonCamera(cameraName);
+        this.cameraToRobotCenter = cameraToRobotCenter;
```
Comment (tuval88) [inline src/main/java/frc/robot/services/poseestimation/PhotonCameraEstimator.java]: This is redundant

File: src/main/java/frc/robot/subsystems/Climber.java [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+ */
+public class Climber extends DoubleSolenoidSubsystem {
+
+    private static final String LEFT_NAMESPACE_NAME = "left climber";
+    private static final String RIGHT_NAMESPACE_NAME = "right climber";
+
+    private static final boolean LEFT_INVERTED = false;
+    private static final boolean RIGHT_INVERTED = false;
+
+    // in 2024, we had 2 climbers; one of each side
```
Comment (tuval88) [inline src/main/java/frc/robot/subsystems/Climber.java]: This is both improper English and a staright-up lie

File: src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 19]
```diff
@@ -1,22 +1,257 @@
...
+
+    public static final Translation2d FRONT_LEFT_WHEEL_POSITION =
+            new Translation2d(TRACK_WIDTH / 2, TRACK_WIDTH / 2);
+    public static final Translation2d FRONT_RIGHT_WHEEL_POSITION =
+            new Translation2d(TRACK_WIDTH / 2, -TRACK_WIDTH / 2);
+    public static final Translation2d BACK_LEFT_WHEEL_POSITION =
+            new Translation2d(-TRACK_WIDTH / 2, TRACK_WIDTH / 2);
+    public static final Translation2d BACK_RIGHT_WHEEL_POSITION =
+            new Translation2d(-TRACK_WIDTH / 2, -TRACK_WIDTH / 2);
+    public static final Translation2d CENTER_OF_ROBOT = new Translation2d(
```
Comment (tuval88) [inline src/main/java/frc/robot/subsystems/Drivetrain.java]: Make this (0,0), and move it to above the wheel positions

File: src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 20]
```diff
@@ -1,22 +1,257 @@
...
+
+    private final SpikesLogger logger = new SpikesLogger("drivetrain");
+
+    private static Drivetrain instance;
+
+    public static Drivetrain getInstance() {
+        if (instance == null) {
+            instance = new Drivetrain(SwerveModuleHolder.getFrontLeft(), SwerveModuleHolder.getFrontRight(),
+                    SwerveModuleHolder.getBackLeft(),
+                    SwerveModuleHolder.getBackRight(), new AHRS(SerialPort.Port.kMXP));
```
Comment (tuval88) [inline src/main/java/frc/robot/subsystems/Drivetrain.java]: Reformat this to have both swerve modules on the same line (and the gyro if there's enough space)

File: src/main/java/frc/robot/subsystems/Drivetrain.java [hunk 21]
```diff
@@ -1,22 +1,257 @@
...
+        Supplier<Double> setpoint = namespace.addConstantDouble("setpoint", 0);
+        namespace.putData("drive in speed i tell you", new DriveSwerve(this, setpoint, () -> 0.0, () -> 0.0, false,
+                true));
+        namespace.putNumber("rotation speed", gyro::getRawGyroZ);
+        namespace.putData("test pathplanner", new InstantCommand(() -> odometry.resetPosition(gyro.getRotation2d(),
+                getModulePositions(), new Pose2d(0, 2, new Rotation2d()))).andThen(
+                new PathPlannerTest(PathPlannerPath.fromPathFile("lol"), this)));
     }
-}
+}
```
Comment (tuval88) [inline src/main/java/frc/robot/subsystems/Drivetrain.java]: Empty line missing

File: src/main/java/frc/robot/subsystems/IntakePlacer.java [hunk 22]
```diff
@@ -0,0 +1,84 @@
...
+    }
+
+    private IntakePlacer(CANSparkMax left, CANSparkMax right) {
+        super(NAMESPACE_NAME, left, right);
+        left.restoreFactoryDefaults();
+        right.restoreFactoryDefaults();
+        left.setIdleMode(CANSparkBase.IdleMode.kBrake);
+        right.setIdleMode(CANSparkBase.IdleMode.kBrake);
+        // left is master, right is slaves.get(0)
+        slaves.get(0).follow(master, true);
```
Comment (tuval88) [inline src/main/java/frc/robot/subsystems/IntakePlacer.java]: This is redundant

File: src/main/java/frc/robot/subsystems/ShooterAdjuster.java [hunk 23]
```diff
@@ -1,17 +1,152 @@
...
+        return feedForwardSettings;
+    }
+
+    // since the subsystem did not have any sensors, we created this method which moves the shooter up until it
+    // surpasses its current limit, in other words it has difficulties moving. the shooter would usually face those
+    // difficulties when it reaches the screw's top, and that way we know that its at the height limit.
+    // this command is supposed to run before using any command which uses the adjuster.
+    public Command getResetCommand() {
+        return new FunctionalCommand(() -> {
+        }, () -> master.set(RESET_SPEED), b -> {
```
Comment (tuval88) [inline src/main/java/frc/robot/subsystems/ShooterAdjuster.java]: rename `b` to `interrupted`

File: src/main/java/frc/robot/subsystems/Storage.java [hunk 24]
```diff
@@ -1,17 +1,76 @@
...
+        sparkMax.setPeriodicFramePeriod(CANSparkLowLevel.PeriodicFrame.kStatus1, STATUS_1_PERIODIC_FRAME);
+        sparkMax.setPeriodicFramePeriod(CANSparkLowLevel.PeriodicFrame.kStatus2, STATUS_2_PERIODIC_FRAME);
+        configureDashboard();
+    }
+
+    public boolean seesNote() {
+        return !infrared.get();
+    }
+
+    public boolean cantMove() {
```
Comment (tuval88) [inline src/main/java/frc/robot/subsystems/Storage.java]: rename to `cannotMove`

**Review by tuval88 — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/commands/Adjust.java (hunk 10 above)
Comment (TurtleMinecraft) [reply src/main/java/frc/robot/commands/Adjust.java:29]: intellij formatting

File: src/main/java/frc/robot/commands/Shoot.java (hunk 13 above)
Comment (TurtleMinecraft) [reply src/main/java/frc/robot/commands/Shoot.java:91]: yes

File: src/main/java/frc/robot/commands/auto/YeetAndRetreatAmpSide.java (hunk 15 above)
Comment (TurtleMinecraft) [reply src/main/java/frc/robot/commands/auto/YeetAndRetreatAmpSide.java]: does it matter?

File: src/main/java/frc/robot/services/poseestimation/PhotonCameraEstimator.java (hunk 17 above)
Comment (TurtleMinecraft) [reply src/main/java/frc/robot/services/poseestimation/PhotonCameraEstimator.java]: how so?

File: src/main/java/frc/robot/commands/auto/YeetAndRetreatAmpSide.java (hunk 15 above)
Comment (tuval88) [reply src/main/java/frc/robot/commands/auto/YeetAndRetreatAmpSide.java]: It means you don't have to multiply its output by 1.0 when you use it, so you can replace STRAFE_SPEED with a method reference

File: src/main/java/frc/robot/services/poseestimation/PhotonCameraEstimator.java (hunk 17 above)
Comment (tuval88) [reply src/main/java/frc/robot/services/poseestimation/PhotonCameraEstimator.java]: This already happens in the super() call

File: src/main/java/frc/robot/subsystems/Climber.java [hunk 25]
```diff
@@ -16,7 +16,7 @@ public class Climber extends DoubleSolenoidSubsystem {
     private static final boolean LEFT_INVERTED = false;
     private static final boolean RIGHT_INVERTED = false;
 
-    // in 2024, we had 2 climbers; one of each side
+    // the robot was supposed to have 2 climbers, one of each side
```
Comment (tuval88) [inline src/main/java/frc/robot/subsystems/Climber.java]: 'one on each side' instead of 'one of each side'

**Review by tuval88 — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/subsystems/Climber.java (hunk 25 above)
Comment (tuval88) [reply src/main/java/frc/robot/subsystems/Climber.java]: Once you fix this, you can consider this PR approved

**Review by Spikes-2212 — APPROVED** [review]

**Review by Spikes-2212 — APPROVED** [review]

**Review by tuval88 — APPROVED** [review]
